docs-src/references/dynamic-configuration.md

```diff
@@ -0,0 +1,173 @@
+---
+id: dynamic-configuration
+title: Dynamic configuration reference
+description: Dynamic condifiguration key values can be set to override the default values in a Cluster configuration.
+sidebar_label: Dynamic configuration
+tags:
+  - reference
+---
+
+Temporal Cluster provides [dynamic configuration](/concepts/what-is-cluster-configuration#dynamic-configuration) keys that you can update and apply to a running Cluster without restarting your services.
+
+The dynamic configuration keys are set with default values when you create your Cluster configuration.
+You can override these values as you test your Cluster setup for optimal performance according to your workload requirements.
+
+For the complete list of dynamic configuration keys, see <https://github.com/temporalio/temporal/blob/master/common/dynamicconfig/constants.go>.
+Ensure that you check server releases notes for any changes to these keys and values.
+
+To check the default values set for a dynamic configuration key, check the following links:
+
+- [Frontend Service](https://github.com/temporalio/temporal/blob/5783e781504d8ffac59f9848b830868f3139b980/service/frontend/service.go#L176)
+- [History Service](https://github.com/temporalio/temporal/blob/5783e781504d8ffac59f9848b830868f3139b980/service/history/configs/config.go#L309)
+- [Matching Service](https://github.com/temporalio/temporal/blob/5783e781504d8ffac59f9848b830868f3139b980/service/matching/config.go#L125)
+- [Worker Service](https://github.com/temporalio/temporal/blob/5783e781504d8ffac59f9848b830868f3139b980/service/worker/service.go#L193)
+
+Note that setting dynamic configuration is optional.
+Change these values only if you need to override the default values to achieve better performance on your Temporal Cluster.
+Also, ensure that you test your changes before setting these in production.
+
+To override the default dynamic configuration values, specify your custom values and constraints for the dynamic configuration keys that you want to change in a YAML configuration file.
+
+- Each dyanmic configuration key can have zero or more values.
+- Each value can have zero or more constraints.
+- There are only three types of constraints that you can define:
+  - `namespace`: `string`
+  - `taskQueueName`: `string`
+  - `taskType`: `int` (`1`:`Workflow`, `2`:`Activity`) A value is selected and returned if all its has exactly the same constraints as the ones specified in query filters (including the number of constraints).
```

Comment:
if value has no constraints, then it's considered as the default value and will be used if there's no exact match.

Reply:
actually, even if number of constraints doesn't match, a value could still be used. e.g. filter is namespace, taskQueue and taskQueueType, if a value only has namespace constraint it can also be selected if there's no better matches.

Check the precedence defined here: https://github.com/temporalio/temporal/blob/378e1e60fbc9761826a7a348ff3f7d1de6284eb0/common/dynamicconfig/collection.go#L476

cc @dnr Please correct me if I am wrong.

Reply:
sorta... the constraints do have the match the _query_ exactly, but the confusing part is that one lookup may use multiple queries. the exact logic is a little involved.

I think somewhere we need to say that each key has a set of allowable constraints. the most common ones are:
- no constraints (it's just global)
- namespace only
- namespace, taskqueuename, tasktype

i.e. it is not the case that any key can be customized by namespace, only some of them can be.

(see comment above, I wrote out how I think it should be explained)